File: src/main/java/chapter4/transporm/SwitchMapExample.java
package chapter4.transporm;

import common.CommonUtils;
import common.Log;
import io.reactivex.Observable;

import java.util.concurrent.TimeUnit;

public class SwitchMapExample {

    public void marbleDiagram() {
        CommonUtils.exampleStart();
        String[] balls = {"1", "3", "5"};
        Observable<String> source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map(Long::intValue)
                .map(idx -> balls[idx])
                .take(balls.length)
                .switchMap(ball -> Observable.interval(200L, TimeUnit.MILLISECONDS)
                        .map(notUsed -> ball + "<>")
                        .take(2)
                );
        source.subscribe(Log::it);
        CommonUtils.sleep(2000);
    }

    public void usingDoOnNext() {
        CommonUtils.exampleStart();
        String[] balls = {"1", "3", "5"};
        Observable<String> source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map(Long::intValue)
                .map(idx -> balls[idx])
                .take(balls.length)
                .doOnNext(Log::it)
                .switchMap(ball -> Observable.interval(200L, TimeUnit.MILLISECONDS)
                        .map(notUsed -> ball + "<>")
                        .take(2)
                );
        source.subscribe(Log::it);
        CommonUtils.sleep(2000);
    }

    public static void main(String[] args) {
        SwitchMapExample demo = new SwitchMapExample();
        demo.marbleDiagram();
        demo.usingDoOnNext();
    }

}

File: src/main/kotlin/chapter4/transporm/SwitchMapExample_.kt
package chapter4.transporm

import common.CommonUtils_
import common.Log_
import io.reactivex.Observable
import java.util.concurrent.TimeUnit


class SwitchMapExample_ {

    fun marbleDiagram() {
        CommonUtils_.exampleStart()
        val balls = arrayOf("1", "3", "5")
        val source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map { it.toInt() }
                .map { balls[it] }
                .take(balls.size.toLong())
                .switchMap { ball ->
                    Observable.interval(200L, TimeUnit.MILLISECONDS)
                            .map { "$ball<>" }
                            .take(2)
                }
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(2000)
    }

    fun usingDoOnNext() {
        CommonUtils_.exampleStart()
        val balls = arrayOf("1", "3", "5")
        val source = Observable.interval(100L, TimeUnit.MILLISECONDS)
                .map { it.toInt() }
                .map { balls[it] }
                .take(balls.size.toLong())
                .doOnNext { Log_.it(it) }
                .switchMap { ball ->
                    Observable.interval(200L, TimeUnit.MILLISECONDS)
                            .map { "$ball<>" }
                            .take(2)
                }
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(2000)
    }

}


fun main(args: Array<String>) {
    val demo = SwitchMapExample_()
    demo.marbleDiagram()
    demo.usingDoOnNext()
}
